[LLD] Extend special OpenBSD support, but scope under ELFOSABI

Ericson2314 opened the pull request:
- Add support for `.openbsd.mutable`

  (rebaser's note) adapted from:
  https://github.com/openbsd/src/commit/bd249b5664da50f0178adea78250a7a0d8ea6566
  New auto-coalescing sections removed

  In the linkers, collect objects in section "openbsd.mutable" and place
  them into a page-aligned region in the bss, with the right markers for
  kernel/ld.so to identify the region and skip making it immutable. While
  here, fix readelf/objdump versions to show all of this. ok miod kettenis

- Add support for `.openbsd.syscalls`

  (rebaser's note) adapted from:
  https://github.com/openbsd/src/commit/42a61acefa8b3288ff2163fb55e934a3fee39974

  Collect .openbsd.syscalls sections into a new PT_OPENBSD_SYSCALLS
  segment. This will be used soon to pin system calls to designated call
  sites.

  ok deraadt@

- Scope OpenBSD special section handling under that ELFOSABI

  As a preexisting comment in `ELF/Writer.cpp` says:

  > section names shouldn't be significant in ELF in spirit.

  so scoping OSABI-specific magic name hacks to just the OSABI in
  question limits the degree to which we deviate from that "spirit" for
  all other OSABIs.

  OpenBSD in particular is very fast moving, having added a number of
  special sections, etc. in recent years. It is unclear how possible /
  reasonable it is for upstream to implement all these features in any
  event, but scoping like this at least mitigates the fallout for other
  OSABIs systems which wish to be more slow-moving.

Depends on #98158

Commit:
[LLD] Extend special OpenBSD support, but scope under ELFOSABI

- Add support for `.openbsd.mutable`

  (rebaser's note) adapted from:
  https://github.com/openbsd/src/commit/bd249b5664da50f0178adea78250a7a0d8ea6566
  New auto-coalescing sections removed

  In the linkers, collect objects in section "openbsd.mutable" and place
  them into a page-aligned region in the bss, with the right markers for
  kernel/ld.so to identify the region and skip making it immutable. While
  here, fix readelf/objdump versions to show all of this. ok miod kettenis

- Add support for `.openbsd.syscalls`

  (rebaser's note) adapted from:
  https://github.com/openbsd/src/commit/42a61acefa8b3288ff2163fb55e934a3fee39974

  Collect .openbsd.syscalls sections into a new PT_OPENBSD_SYSCALLS
  segment. This will be used soon to pin system calls to designated call
  sites.

  ok deraadt@

- Scope OpenBSD special section handling under that ELFOSABI

  As a preexisting comment in `ELF/Writer.cpp` says:

  > section names shouldn't be significant in ELF in spirit.

  so scoping OSABI-specific magic name hacks to just the OSABI in
  question limits the degree to which we deviate from that "spirit" for
  all other OSABIs.

  OpenBSD in particular is very fast moving, having added a number of
  special sections, etc. in recent years. It is unclear how possible /
  reasonable it is for upstream to implement all these features in any
  event, but scoping like this at least mitigates the fallout for other
  OSABIs systems which wish to be more slow-moving.

Ericson2314 commented:
I'm happy to report that with change, my cross compiled statically linked hello world ran successfully :)

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -107,7 +107,8 @@ static StringRef getOutputSectionName(const InputSectionBase *s) {
         ".lrodata", ".lbss", ".gcc_except_table", ".init_array", ".fini_array",
-        ".tbss", ".tdata", ".ARM.exidx", ".ARM.extab", ".ctors", ".dtors"})
+        ".tbss", ".tdata", ".ARM.exidx", ".ARM.extab", ".ctors", ".dtors",
+        ".openbsd.randomdata", ".openbsd.mutable"})
```
OpenBSD is adding program headers in an alarming rate... (https://isopenbsdsecu.re/mitigations/)

While ad-hoc `PT_OPENBSD_*` are somewhat fine, changing the default linker script would add overhead for other OSes. Does OpenBSD really need this? I fairly strongly oppose to do this part of change.

Ericson2314 replied:
I have not tried to bisect this patch, but I can assure you that with out my "hello world' did not run (because malloc in OpenBSD's libc uses `.openbsd.mutable` for the root object of the heap, and there was an issue initializing it).

OpenBSD has a much faster-moving ABI, and I agree we wouldn't want that intentional stability to affect other platforms that are not opting into it. This is why I lean towards gating all these things under `config->osabi == ELFOSABI_FREEBSD` --- OpenBSD therefore has its own sandbox where it is sure not to affect other platforms. I just don't know how to set `config->osbi` from the command-line in a satisfactory way, hence https://github.com/llvm/llvm-project/issues/97124

Ericson2314 replied:
OK I did a test and this is unnecessary for my cross-compiled statically-linked binaries. Removed!

Ericson2314 commented:
@MaskRay After testing that my cross compiled binary still runs, I:

- Removed the section-coalescing parts of the first two commits.

- Added a new commit which:

  - Gated the `PT_`-assignments under a `config->osabi == ELFOSABI_OPENBSD` check

   - Modified LLVM to set `ELFOSABI_OPENBSD` for OpenBSD targets just like it does `ELFOSABI_FREEBSD` for FreeBSD ones

I am reasonably happy with this. Per the 3rd commits message, without weighing in on the degree to which it is possible/reasonable for upstream to track something as fast-moving as OpenBSD, the `config->osabi == ELFOSABI_OPENBSD` adds a stronger "firewall" between OpenBSD and everything else.

Ericson2314 commented:
I rebased atop #98158 so as to separate the LLVM an LLD parts, in case that is easier for review.

MaskRay commented on lld/ELF/ScriptParser.cpp:
```diff
@@ -1628,7 +1628,9 @@ unsigned ScriptParser::readPhdrType() {
                      .Case("PT_GNU_EH_FRAME", PT_GNU_EH_FRAME)
                      .Case("PT_GNU_STACK", PT_GNU_STACK)
                      .Case("PT_GNU_RELRO", PT_GNU_RELRO)
+                     .Case("PT_OPENBSD_MUTABLE", PT_OPENBSD_MUTABLE)
```
Add tests in ELF/openbsd-phdr.s

MaskRay commented on llvm/include/llvm/MC/MCELFObjectWriter.h:
```diff
@@ -78,6 +78,8 @@ class MCELFObjectTargetWriter : public MCObjectTargetWriter {
         return ELF::ELFOSABI_FREEBSD;
       case Triple::Solaris:
         return ELF::ELFOSABI_SOLARIS;
+      case Triple::OpenBSD:
```
This should be separate with a test https://reviews.llvm.org/D107748

Ericson2314 replied:
Thanks @MaskRay, #98158 has just that, with such tests

MaskRay commented on lld/ELF/Writer.cpp:
```diff
@@ -2273,10 +2279,26 @@ SmallVector<PhdrEntry *, 0> Writer<ELFT>::createPhdrs(Partition &part) {
-    addHdr(PT_OPENBSD_RANDOMIZE, cmd->getPhdrFlags())->add(cmd);
+  // Handle OpenBSD-specific section types for OpenBSD "OS ABI".
+  //
+  // Scoped to just this "OS ABI" because, as is written below "section
```
The comment "as is written .. " can be deleted. It's clear that `if (config->osabi == ELFOSABI_OPENBSD)` doesn't apply to other OSes.

Ericson2314 commented:
OK @MaskRay I think this is ready now?

- (Previously) addressed your 3 prior review comments
- Rebased atop merged #98158 so it's just the one commit
- changed `-triple` to be `x86_64-unknown-freebsd` (with `unknown`) to match what you did in 06cecdc60ec9ebfdd4d8cdb2586d201272bdf6bd

Ericson2314 commented:
(Rebased because one thing in https://github.com/llvm/llvm-project/pull/98158 needed fixing.)

Ericson2314 commented:
I found another such possibly-affected test. I need to make sure it is working --- even if the section rename is not necessary, I figure either keeping it, or making the target `openbsd` not `linux` is a good idea.

MaskRay commented on lld/ELF/Writer.cpp:
```diff
@@ -601,10 +601,16 @@ static bool isRelroSection(const OutputSection *sec) {
-         s == ".fini_array" || s == ".init_array" ||
-         s == ".openbsd.randomdata" || s == ".preinit_array";
+
+  bool abiAgnostic = s == ".data.rel.ro" || s == ".bss.rel.ro" ||
```
It should be fine to keep this unchanged. We can add the osabi check if the `.openbsd.*` list ever gets longer.

Ericson2314 replied:
(And checked locally that both the new tests worked, and the old test failed.)

MaskRay commented on lld/ELF/Writer.cpp:
```diff
@@ -2273,10 +2279,23 @@ SmallVector<PhdrEntry *, 0> Writer<ELFT>::createPhdrs(Partition &part) {
-  // the dynamic linker fill the segment with random data.
-  if (OutputSection *cmd = findSection(".openbsd.randomdata", partNo))
-    addHdr(PT_OPENBSD_RANDOMIZE, cmd->getPhdrFlags())->add(cmd);
+  // Handle OpenBSD-specific section types for OpenBSD "OS ABI".
```
This comment can be removed. the code self explains

Ericson2314 replied:
Done

Ericson2314 commented:
Linux CI is stuck in the queue, Windows has
```
PASS: lld :: ELF/openbsd-phdr.s (1928 of 2949)
...
PASS: lld :: ELF/relro-openbsd.s (2129 of 2949)
PASS: lld :: ELF/relro.s (2130 of 2949)
```
and there is no thing host-specific about this, so I think we are good!
